base/compiler/ssair/inlining.jl

```diff
@@ -1073,9 +1076,12 @@ function analyze_single_call!(ir::IRCode, todo::Vector{Pair{Int, Any}}, idx::Int
     for i in 1:length(infos)
         info = infos[i]
         meth = info.results
-        if meth === missing || meth.ambig
+        if meth === missing
             # Too many applicable methods
-            # Or there is a (partial?) ambiguity
+            too_many = true
+            break
+        elseif meth.ambig
+            # there is a (partial?) ambiguity
             too_many = true
```

Comment:
Constraints don't propagate well in the following case, somehow:
```julia
code_typed((Union{Int,Nothing},)) do x
    if isa(x, Nothing) || x < 0
        return 0
    end
    x # x::Int, ideally
end |> first
```

I will investigate into this when I have time.

Reply:
It is the lowering structure. For `&&`, it is lowered to a nested `if` block, but not so for `||`. So for entry to `#4`, we have the tmerge for `[x = Nothing, _3 = Conditional(:x, Nothing, Int)]` and `[x = Int, _3 = Bool]`. Instead, lowering could fuse the `goto #4` into the next line (jump-threading), and do `goto #5` there to fix this. Or we should narrow the later `_3 = Bool` to the equivalent `_3 = Conditional(x, Int, Int)` before the corresponding tmerge. Either way, a good update to consider for later:
```
julia> code_typed((Union{Int,Nothing},), optimize=false) do x
           if x === nothing || x < 0
               return 0
           end
           x # x::Int, ideally
       end
1-element Vector{Any}:
 CodeInfo(
    @ REPL[6]:2 within `#13'
1 ─ %1 = (x === Main.nothing)::Bool
└──      goto #3 if not %1
2 ─      (@_3 = %1)::Bool
└──      goto #4
3 ─      (@_3 = x::Int64 < 0)::Bool
4 ┄      goto #6 if not @_3
    @ REPL[6]:3 within `#13'
5 ─      return 0
    @ REPL[6]:5 within `#13'
6 ─      return x
) => Union{Nothing, Int64}
```

Reply:
yeah, filed as the issue for now: https://github.com/JuliaLang/julia/issues/39611

Reply:
This part is reverted back to the original code. Thanks @simeonschaub !